controllers/topology/internal/extensions/patches/api/interface.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package api contains the API definition for the patch engine.
+package api
+
+import (
+	"context"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+)
+
+// Generator defines a component that can generate patches for ClusterClass templates.
+// NOTE: We are introducing this interface as a decoupling layer between the patch engine and the concrete components
+// responsible for generating patches, because we aim to provide support for external patches in a future iteration.
+// We also assume that this interface and all the related types will be moved in a separated (versioned) package thus
+// providing a versioned contract between Cluster API and the components implementing external patch extensions.
+type Generator interface {
+	// Generate generates patches for templates.
+	// GenerateRequest contains templates and the corresponding variables.
+	// GenerateResponse contains the patches which should be applied to the templates of the GenerateRequest.
+	Generate(ctx context.Context, request *GenerateRequest) (*GenerateResponse, error)
+}
+
+// GenerateRequest defines the input for a Generate request.
+type GenerateRequest struct {
+	// Variables is a name/value map containing variables.
+	Variables map[string]apiextensionsv1.JSON
+
+	// Items contains the list of templates to generate patches for.
+	Items []*GenerateRequestTemplate
+}
+
+// GenerateRequestTemplate defines one of the ClusterClass templates to generate patches for.
+type GenerateRequestTemplate struct {
+	// TemplateID identifies a template to generate patches for;
+	// the same TemplateID must be used when specifying to which template a generated patch should be applied to.
+	TemplateID TemplateID
+
+	// Variables is a name/value map containing variables specifically for the current template.
+	// For example some builtin variables like MachineDeployment replicas and version are context-sensitive
+	// and thus are only added to templates for MachineDeployments and with values which correspond to the
+	// current MachineDeployment.
+	Variables map[string]apiextensionsv1.JSON
+
+	// Template contains the template.
+	Template apiextensionsv1.JSON
+}
+
+// TemplateID identifies one of the ClusterClass templates to generate patches for;
+// the same TemplateID must be used when specifying where a generated patch should apply to.
+type TemplateID struct {
```

Comment:
This struct, and the `MachineDeploymentID` one have the suffix `ID` which could be confusing — Could we call them something different? Left a few suggestions here and there

Reply:
Renamed to TemplateRef and MachineDeploymentRef